subprojects/build-operations/src/main/java/org/gradle/internal/operations/CurrentBuildOperationRef.java

```diff
@@ -56,6 +56,23 @@ public void clear() {
         ref.remove();
     }
 
+    /**
+     * Callable with generic exception.
+     */
+    public interface Callable<T, E extends Throwable> {
+        T call() throws E;
+    }
+
+    public <T, E extends Throwable> T with(@Nullable BuildOperationRef state, Callable<T, E> block) throws E {
```

Comment:
Would a regular Callable not work here?

Reply:
Technically I could do it that way, but it would make the call in `WorkerDaemonFactory` awkward as I would need to handle a checked exception that would never get thrown. I would prefer to keep it this way to avoid weird handling code elsewhere.

Reply:
Could this be a `Factory` in that case?

Reply:
No, it also produces `IOException` inside `ExecOutputHandleRunner#forwardContent`.

Reply:
Though, to be fair, that should probably use the `Runnable` variant, and I could add a custom version of that, and remove the `Callable`.